src/core/border_router/nat64_translator.cpp

```diff
@@ -0,0 +1,454 @@
+
+    mapping = CreateMapping(aAddr);
+
+exit:
+    return mapping;
+}
+
+Nat64Translator::AddressMapping *Nat64Translator::GetMapping(const Ip4::Address &aAddr)
+{
+    AddressMapping *mapping = mActiveAddressMappings.FindMatching(aAddr);
+
+    if (mapping != nullptr)
+    {
+        mapping->Touch(Get<Uptime>().GetUptime());
+    }
+    return mapping;
+}
+
+Error Nat64Translator::TranslateIcmp4(AddressMapping &aMapping, Message &aMessage)
+{
+    Error             err = kErrorNone;
+    Ip4::Icmp::Header icmp4Header;
+    Ip6::Icmp::Header icmp6Header;
+
+    OT_UNUSED_VARIABLE(aMapping);
+
+    VerifyOrExit(aMessage.ReadBytes(0, &icmp4Header, sizeof(icmp4Header)) == sizeof(icmp4Header), err = kErrorParse);
+    switch (icmp4Header.GetType())
+    {
+    case Ip4::Icmp::Header::Type::kTypeEchoReply:
+    {
+        VerifyOrExit(aMessage.ReadBytes(0, &icmp6Header, sizeof(icmp6Header)) == sizeof(icmp6Header),
+                     err = kErrorParse);
+        icmp6Header.SetType(Ip6::Icmp::Header::Type::kTypeEchoReply);
+        aMessage.WriteBytes(0, &icmp6Header, sizeof(icmp6Header));
+        break;
+    }
+    default:
+        err = kErrorInvalidArgs;
+        break;
+    }
+
+exit:
+    return err;
+}
+
+Error Nat64Translator::TranslateIcmp6(AddressMapping &aMapping, Message &aMessage)
+{
+    Error             err = kErrorNone;
+    Ip4::Icmp::Header icmp4Header;
+    Ip6::Icmp::Header icmp6Header;
+
+    OT_UNUSED_VARIABLE(aMapping);
+
+    VerifyOrExit(aMessage.ReadBytes(0, &icmp6Header, sizeof(icmp6Header)) == sizeof(icmp6Header), err = kErrorParse);
+    switch (icmp6Header.GetType())
+    {
+    case Ip6::Icmp::Header::Type::kTypeEchoRequest:
+    {
+        VerifyOrExit(aMessage.ReadBytes(0, &icmp4Header, sizeof(icmp4Header)) == sizeof(icmp4Header),
+                     err = kErrorParse);
+        icmp4Header.SetType(Ip4::Icmp::Header::Type::kTypeEchoRequest);
+        aMessage.WriteBytes(0, &icmp4Header, sizeof(icmp4Header));
+        break;
+    }
+    default:
+        err = kErrorInvalidArgs;
+        break;
+    }
+
+exit:
+    return err;
+}
+
+Error Nat64Translator::SetIp4Cidr(const Ip4::Cidr &aCidr)
+{
+    Error err = kErrorNone;
+
+    uint32_t numberOfHosts;
+    uint32_t hostIdBegin;
+
+    VerifyOrExit(aCidr.mLength > 0 && aCidr.mLength <= 32, err = kErrorInvalidArgs);
+
+    VerifyOrExit(mIp4Cidr != aCidr);
+
+    // Avoid using the 0s and 1s in the host id of an address, but what if the user provides us with /32 or /31
+    // addresses?
+    if (aCidr.mLength == 32)
+    {
+        hostIdBegin   = 0;
+        numberOfHosts = 1;
+    }
+    else if (aCidr.mLength == 31)
+    {
+        hostIdBegin   = 0;
+        numberOfHosts = 2;
+    }
+    else
+    {
+        hostIdBegin   = 1;
+        numberOfHosts = static_cast<uint32_t>((1 << (Ip4::Address::kSize * 8 - aCidr.mLength)) - 2);
+    }
+    numberOfHosts = OT_MIN(numberOfHosts, kAddressMappingPoolSize);
+
+    mAddressMappingPool.FreeAll();
```

Comment:
Should also clear the `mActiveAddressMappings` linked list?
(in case some entries from pool were already allocated from the pool and are in the linked list).

Reply:
True, we should release mActiveAddressMappings.